src/algorithms/pso_gen.cpp

```diff
@@ -362,26 +364,76 @@ population pso_gen::evolve(population pop) const
             pop.get_problem().set_seed(urng(m_e));
             // re-evaluate the whole population w.r.t. the new seed
 
-            for (decltype(swarm_size) p = 0u; p < swarm_size; ++p) {
-                // We evaluate here the new individual fitness
-                fit[p] = prob.fitness(X[p]);
-                // We re-evaluate the fitness of the particle memory
-                lbfit[p] = prob.fitness(lbX[p]);
-            }
+            if (m_bfe) {
+                // bfe is available:
+                vector_double decision_vectors(swarm_size * dim);
+                vector_double decision_vectors_lb(swarm_size * dim);
 
-            best_fit = fit[0];
-            best_neighb = X[0];
+                for (decltype(swarm_size) i = 0; i < swarm_size; ++i) {
+                    // I store the individuals in a contiguous vector
+                    std::copy(X[i].begin(), X[i].end(), decision_vectors.data() + i * dim);
+                    std::copy(lbX[i].begin(), lbX[i].end(), decision_vectors_lb.data() + i * dim);
+                }
 
-            for (decltype(swarm_size) p = 1; p < swarm_size; p++) {
-                if (fit[p] < best_fit) {
-                    best_fit = fit[p];
-                    best_neighb = X[p];
+                auto fitnesses = (*m_bfe)(prob, decision_vectors);
+                auto fitnesses_lb = (*m_bfe)(prob, decision_vectors_lb);
+                for (decltype(swarm_size) i = 0; i < swarm_size; ++i) {
+                    for (decltype(fit[i].size()) i_fit = 0; i_fit < fit[i].size(); ++i_fit) {
+                        fit[i][i_fit] = fitnesses[i];
+                        lbfit[i][i_fit] = fitnesses_lb[i];
+                    }
+                }
+                best_fit = fit[0];
+                best_neighb = X[0];
+
+                for (decltype(swarm_size) p = 1; p < swarm_size; p++) {
+                    if (fit[p] < best_fit) {
```

Comment:
@darioizzo these fitness comparisons should be done using ``less_than_f`` right?

Reply:
Yeps, fprgot about it .... its a pattern that is to be fixed in many other UDAs. @Sceki When comparing a fitness value to another using the ```<``` operator, the result is basically subject to issues in cases where the fitness is ```nan``` (happes if the UDP ihas numerical instabilities and returns nan). In C++ any comparison agains nan fails. So ```0.3 < nan``` is false and you loose a good individual.

To avoid this, we have  in ```detail\custom_comparisons.hpp``` a helper function called ```less_than_f ``` that places ```nans``` after ```inf```. so that ```3<inf<nan```. Can you use it here throughout?

btw also in gaco and any other UDA you happen to work on .....

Reply:
Ok, I will do it!

I think I have used it for the algos I implemented as we already discussed this in the past..